### PR TITLE
Rewrite ddim

### PR DESCRIPTION
**Pull request by sneaxiy**
test=develop

### CODE REVIEW

File: paddle/fluid/framework/ddim.h [hunk 1]
```diff
@@ -18,62 +18,136 @@ limitations under the License. */
...
+  inline int64_t* GetMutable() { return dim_.GetMutable(); }
 
-  DDim operator+(DDim d) const;
+  inline int size() const { return rank_; }
 
-  DDim operator*(DDim d) const;
+ private:
+  template <int D>
+  inline Dim<D>& UnsafeCast() {
+    return const_cast<Dim<D>&>(const_cast<const DDim*>(this)->UnsafeCast<D>());
```
Comment (reyoung) [inline paddle/fluid/framework/ddim.h]: Perhaps we can avoid const_cast by rewrite UnsafeCast twice.

File: paddle/fluid/framework/dim.h [hunk 2]
```diff
@@ -16,332 +16,184 @@
...
+}  // namespace detail
 
-///\cond HIDDEN
-// Base case of ex_prefix_mul
-// Notice it is inline because it is no longer a template
-template <>
-HOSTDEVICE inline Dim<0> ex_prefix_mul(const Dim<0>& src, int mul) {
-  return Dim<0>();
+template <int D>
+HOSTDEVICE inline Dim<D> ex_prefix_mul(const Dim<D>& src) {
```
Comment (reyoung) [inline paddle/fluid/framework/dim.h:164]: It seems that a lot of methods in this file are not used by other modules. Perhaps we can shrink them later.

**Review by reyoung — DISMISSED** [review]
LGTM

However, some APIs are not useful, we can delete them later.

File: paddle/fluid/framework/ddim.h (hunk 1 above)
Comment (sneaxiy) [reply paddle/fluid/framework/ddim.h]: Of course. I would rewrite it.

File: paddle/fluid/framework/dim.h (hunk 2 above)
Comment (sneaxiy) [reply paddle/fluid/framework/dim.h:164]: Yes. Maybe in next PR cleaning the unused codes.

File: paddle/fluid/framework/ddim.h (hunk 1 above)
Comment (sneaxiy) [reply paddle/fluid/framework/ddim.h]: Done.

**Review by reyoung — APPROVED** [review]

**Comment by zhiqiu** [comment]
Hi, @sneaxiy I read the code of `ddim` recently, and got a little confused. 
 Would you please explain some of the background and motivation of rewriting ddim?  Thanks.

**Pull request merged**